File: timeline/src/profiles/real.ts
import {Profile, ProfileType, Windows} from '../internal.js';
import {Segment} from '../segment.js';
import type {Timeline} from '../timeline.js';
import {coalesce} from '../timeline.js';
import {Inclusivity, Interval} from '../interval.js';
import {BinaryOperation} from '../binary-operation.js';
import {Temporal} from '@js-temporal/polyfill';
import { fetcher } from "../data-fetcher.js";

// @ts-ignore
export class Real extends Profile<LinearEquation> {
  constructor(segments: Timeline<Segment<LinearEquation>>) {
    super(segments, ProfileType.Real);
  }

  public static empty(): Real {
    return new Real(async _ => []);
  }

  public static override Value(value: number, interval?: Interval): Real;
  public static override Value(value: LinearEquation, interval?: Interval): Real;
  public static override Value(value: number | LinearEquation, interval?: Interval): Real {
    if (typeof value === 'number') value = LinearEquation.Constant(value);
    return new Real(async bounds => [
      new Segment(value as LinearEquation, interval === undefined ? bounds : Interval.intersect(bounds, interval))
    ]);
  }

  public static override Resource(name: string): Real {
    return new Real(fetcher.resource(name, (v, t) => new LinearEquation(t, v.initial as number, v.rate as number), ProfileType.Real));
  }

  public override assignGaps(def: Profile<LinearEquation> | LinearEquation | number): Real {
    if (typeof def === 'number') def = LinearEquation.Constant(def);
    return super.assignGaps(def);
  }

  public negate(): Real {
    return this.mapValues(v => v.negate());
  }

  public abs(): Real {
    return this.unsafe.flatMap(
      (eq, i) => eq.abs(i),
      b => b
    );
  }

  public plus(other: number | Profile<LinearEquation>): Real {
    if (typeof other === 'number') other = Real.Value(other);
    return this.map2Values(
      other,
      BinaryOperation.combineOrUndefined((l, r) => l.plus(r))
    );
  }

  public minus(other: number | Profile<LinearEquation>): Real {
    if (typeof other === 'number') other = Real.Value(other);
    return this.map2Values(
      other,
      BinaryOperation.combineOrUndefined((l, r) => l.minus(r))
    );
  }

  public times(other: number | Profile<LinearEquation>): Real {
    if (typeof other === 'number') return this.mapValues(eq => eq.times(other));
    else
      return this.map2Values(
        other,
        BinaryOperation.combineOrUndefined((l, r, i) => {
          if (l.rate === 0) {
            return new LinearEquation(r.initialTime, l.initialValue * r.initialValue, l.initialValue * r.rate);
          } else if (r.rate === 0) {
            return new LinearEquation(l.initialTime, r.initialValue * l.initialValue, r.initialValue * l.rate);
          }
          throw new Error(
            `Cannot multiply two non-constant real profiles, rates were ${l.rate} and ${r.rate}. Interval: ${i}`
          );
        }),
        ProfileType.Real
      );
  }

  public dividedBy(other: number | Profile<LinearEquation>): Real {
    if (typeof other === 'number') return this.mapValues(eq => eq.dividedBy(other));
    else
      return this.map2Values(
        other,
        BinaryOperation.combineOrUndefined((l, r, i) => {
          if (r.rate !== 0)
            throw new Error(
              `Denominator of real profile division must be piece-wise constant. Segment: ${new Segment(r, i)}`
            );
          return l.dividedBy(r.initialValue);
        }),
        ProfileType.Real
      );
  }

  public pow(exp: number | Profile<LinearEquation>): Real {
    if (typeof exp === 'number') {
      if (exp === 1) return this;
      if (exp === 0) return Real.Value(1);
      exp = Real.Value(exp);
    }
    return this.map2Values(
      exp,
      BinaryOperation.combineOrUndefined((eq, e, i) => {
        if (e.rate !== 0)
          throw new Error(`Cannot raise to an exponent that is not piece-wise constant. Segment: ${new Segment(e, i)}`);
        if (eq.rate !== 0)
          throw new Error(
            `Cannot exponentiate a real profile that is not piece-wise constant. Segment: ${new Segment(eq, i)}`
          );
        const newValue = Math.pow(eq.initialValue, e.initialValue);
        if (isNaN(newValue))
          throw new Error(
            `Exponentiation returned NaN on numbers ${eq.initialValue} ^ ${e.initialValue}; see Math.pow documentation for possible reasons why. Interval: ${i}`
          );
        return new LinearEquation(eq.initialTime, newValue, 0);
      })
    );
  }

  public root(n: number): Real {
    return this.pow(1 / n);
  }

  public sqrt(): Real {
    return this.pow(0.5);
  }

  public rate(unit?: Temporal.Duration): Real {
    if (unit !== undefined) return this.mapValues(v => LinearEquation.Constant(v.rate));
    else return this.mapValues(v => LinearEquation.Constant((v.rate * unit!.total('microsecond')) / 1_000_000));
  }

  public integrate(unit: Temporal.Duration): Real {
    const timeline = async (bounds: Interval) => {
      const baseRate = 1 / unit.total('second');
      const segments = await this.segments(bounds);
      const result: Segment<LinearEquation>[] = [];
      let previousTime = bounds.start;
      let acc = 0;
      for (const segment of segments) {
        if (Temporal.Duration.compare(previousTime, segment.interval.start) !== 0)
          throw new Error(
            `Cannot integrate a real profile with gaps. Gap: ${Interval.Between(previousTime, segment.interval.start)}`
          );
        if (segment.value.rate !== 0)
          throw new Error(`Cannot integrate real profiles that aren't piecewise constant. Segment: ${segment}.`);
        let rate = segment.value.initialValue * baseRate;
        let nextAcc = acc + rate * segment.interval.duration().total('second');
        result.push(new Segment(new LinearEquation(previousTime, acc, rate), segment.interval));
        previousTime = segment.interval.end;
        acc = nextAcc;
      }
      return result;
    };
    return new Real(timeline);
  }

  public compare(other: Profile<LinearEquation>, comparator: (l: number, r: number) => boolean): Windows {
    return this.unsafe.flatMap2(
      other,
      BinaryOperation.combineOrUndefined((l, r, interval) => l.compare(r, comparator, interval)),
      b => b,
      ProfileType.Windows
    );
  }

  public override equalTo(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l === r);
  }

  public override notEqualTo(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l !== r);
  }

  public lessThan(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l < r);
  }

  public lessThanOrEqual(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l <= r);
  }

  public greaterThan(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l > r);
  }

  public greaterThanOrEqual(other: number | Profile<LinearEquation>): Windows {
    if (typeof other === 'number') other = Real.Value(other);
    return this.compare(other, (l, r) => l >= r);
  }

  public override changes(): Windows {
    const segments = async (bounds: Interval) => {
      let previous: Segment<LinearEquation> | undefined = undefined;
      return coalesce(
        (await this.segments(bounds)).flatMap(currentSegment => {
          let leftEdge: boolean | undefined;

          const currentInterval = currentSegment.interval;

          if (previous !== undefined) {
            if (
              Interval.compareEndToStart(previous.interval, currentInterval) === 0 &&
              currentInterval.includesStart()
            ) {
              leftEdge =
                previous.value.valueAt(currentInterval.start) === currentSegment.value.valueAt(currentInterval.start);
            } else {
              leftEdge = undefined;
            }
          } else {
            if (Interval.compareStarts(currentInterval, bounds) === 0) {
              leftEdge = currentSegment.value.rate !== 0;
            } else {
              leftEdge = undefined;
            }
          }

          return [
            new Segment(leftEdge, Interval.At(currentInterval.start)).transpose(),
            new Segment(
              currentSegment.value.rate !== 0,
              Interval.Between(currentInterval.start, currentInterval.end, Inclusivity.Exclusive)
            )
          ].filter($ => $ !== undefined) as Segment<boolean>[];
        }),
        ProfileType.Real
      );
    };
    return new Windows(segments);
  }

  public override transitions(from: number, to: number): Windows;
  public override transitions(from: LinearEquation, to: LinearEquation): Windows;
  public override transitions(from: number | LinearEquation, to: number | LinearEquation): Windows {
    if (typeof from === 'number') {
      return this.edges(
        BinaryOperation.combineOrUndefined((l, r, i) => l.valueAt(i.start) === from && r.valueAt(i.start) === to)
      );
    } else return super.transitions(from, to as LinearEquation);
  }
}

export class LinearEquation {
  public initialTime: Temporal.Duration;
  public initialValue: number;
  public rate: number;

  constructor(initialTime: Temporal.Duration, initialValue: number, rate: number) {
    this.initialTime = initialTime;
    this.initialValue = initialValue;
    this.rate = rate;
  }

  public static Constant(c: number): LinearEquation {
    return new LinearEquation(new Temporal.Duration(), c, 0);
  }

  public negate(): LinearEquation {
    return new LinearEquation(this.initialTime, -this.initialValue, -this.rate);
  }

  public abs(bounds: Interval): Segment<LinearEquation>[] {
    const root = this.intersectionPointWith(LinearEquation.Constant(0));
    if (root === undefined || !bounds.contains(root)) {
      if (this.valueAt(bounds.start) > 0)
        return [new Segment(new LinearEquation(this.initialTime, this.initialValue, this.rate), bounds)];
      else return [new Segment(new LinearEquation(this.initialTime, -this.initialValue, -this.rate), bounds)];
    }
    return [
      new Segment(
        new LinearEquation(root, 0, -Math.abs(this.rate)),
        Interval.Between(bounds.start, root, bounds.startInclusivity, Inclusivity.Exclusive)
      ),
      new Segment(
        new LinearEquation(root, 0, Math.abs(this.rate)),
        Interval.Between(root, bounds.end, Inclusivity.Inclusive, bounds.endInclusivity)
      )
    ];
  }

  public times(c: number): LinearEquation {
    return new LinearEquation(this.initialTime, c * this.initialValue, c * this.rate);
  }

  public dividedBy(c: number): LinearEquation {
    return this.times(1 / c);
  }

  public plus(other: LinearEquation | number): LinearEquation {
    if (typeof other === 'number') other = LinearEquation.Constant(other);
    const shifted = other.shiftInitialTime(this.initialTime);
    return new LinearEquation(this.initialTime, this.initialValue + shifted.initialValue, this.rate + other.rate);
  }

  public minus(other: LinearEquation | number): LinearEquation {
    if (typeof other === 'number') return this.plus(-other);
    return this.plus(other.negate());
  }

  public valueAt(time: Temporal.Duration): number {
    const change = this.rate * time.total('second') - this.rate * this.initialTime.total('second');
    return this.initialValue + change;
  }

  public shiftInitialTime(newInitialTime: Temporal.Duration): LinearEquation {
    return new LinearEquation(
      newInitialTime,
      this.initialValue + newInitialTime.subtract(this.initialTime).total('second') * this.rate,
      this.rate
    );
  }

  public equals(other: LinearEquation): boolean {
    return this.initialValue === other.valueAt(this.initialTime) && this.rate === other.rate;
  }

  public intersectionPointWith(other: LinearEquation): Temporal.Duration | undefined {
    if (this.rate === other.rate) return undefined;
    return this.initialTime.add(
      Temporal.Duration.from({
        microseconds: Math.round(
          (1_000_000 * (other.valueAt(this.initialTime) - this.initialValue)) / (this.rate - other.rate)
        )
      })
    );
  }

  public compare(
    other: LinearEquation,
    comparator: (l: number, r: number) => boolean,
    bounds: Interval
  ): Segment<boolean>[] {
    const intersection = this.intersectionPointWith(other);
    if (intersection === undefined)
      return [new Segment(comparator(this.initialValue, other.valueAt(this.initialTime)), bounds)];
    else {
      return [
        new Segment(
          comparator(this.valueAt(bounds.start), other.valueAt(bounds.start)),
          Interval.Between(bounds.start, intersection, bounds.startInclusivity, Inclusivity.Exclusive)
        ),
        new Segment(comparator(this.valueAt(intersection), other.valueAt(intersection)), Interval.At(intersection)),
        new Segment(
          comparator(this.valueAt(bounds.end), other.valueAt(bounds.end)),
          Interval.Between(intersection, bounds.end, Inclusivity.Exclusive, bounds.endInclusivity)
        )
      ].filter($ => !$.interval.isEmpty());
    }
  }
}

File: timeline/src/main/kotlin/gov/nasa/jpl/aerie/timeline/collections/profiles/Real.kt
package gov.nasa.jpl.aerie.timeline.collections.profiles

import gov.nasa.jpl.aerie.merlin.protocol.types.SerializedValue
import gov.nasa.jpl.aerie.timeline.*
import gov.nasa.jpl.aerie.timeline.ops.numeric.LinearOps
import gov.nasa.jpl.aerie.timeline.payloads.Segment
import gov.nasa.jpl.aerie.timeline.ops.numeric.SerialNumericOps
import gov.nasa.jpl.aerie.timeline.payloads.LinearEquation
import gov.nasa.jpl.aerie.timeline.payloads.transpose
import gov.nasa.jpl.aerie.timeline.util.preprocessList
import gov.nasa.jpl.aerie.timeline.util.truncateList
import kotlin.jvm.optionals.getOrNull
import kotlin.math.pow

/** A profile of [LinearEquations][LinearEquation]; a piece-wise linear real-number profile. */
data class Real(private val timeline: Timeline<Segment<LinearEquation>, Real>):
    Timeline<Segment<LinearEquation>, Real> by timeline,
    SerialNumericOps<LinearEquation, Real>,
    LinearOps<Real>
{
  constructor(v: Int): this(v.toDouble())
  constructor(v: Long): this(v.toDouble())
  constructor(v: Double): this(LinearEquation(v))
  constructor(eq: LinearEquation): this(Segment(Interval.MIN_MAX, eq))
  constructor(vararg segments: Segment<LinearEquation>): this(segments.asList())
  constructor(segments: List<Segment<LinearEquation>>): this(BaseTimeline(::Real, preprocessList(segments, Segment<LinearEquation>::valueEquals)))

  override fun toSerialLinear() = this
  override fun LinearEquation.toLinear() = this

  /**
   * Converts this to a primitive number profile (i.e. [Numbers]), throwing an error if this is not piece-wise constant.
   *
   * @param message error message to throw if this is not piece-wise constant.
   */
  override fun toSerialPrimitiveNumbers(message: String?) = mapValues(::Numbers) {
    if (it.value.isConstant()) it.value.initialValue
    else if (message == null) throw RealOpException("Cannot convert a non-piecewise-constant linear equation to a constant number. (at time ${it.interval.start})")
    else throw RealOpException("$message (at time ${it.interval.start})")
  }

  /** Adds this and another numeric profile. */
  operator fun plus(other: SerialNumericOps<*, *>) = map2Values(other.toSerialLinear()) { l, r, _ ->
    val shiftedRight = r.shiftInitialTime(l.initialTime)
    LinearEquation(l.initialTime, l.initialValue + shiftedRight.initialValue, l.rate + r.rate)
  }

  /** Adds a constant number to this. */
  operator fun plus(n: Number) = plus(Numbers(n))

  /** Subtracts another numeric profile from this. */
  operator fun minus(other: SerialNumericOps<*, *>) = map2Values(other.toSerialLinear()) { l, r, _ ->
    val shiftedRight = r.shiftInitialTime(l.initialTime)
    LinearEquation(l.initialTime, l.initialValue - shiftedRight.initialValue, l.rate - r.rate)
  }

  /** Subtracts a constant number from this. */
  operator fun minus(n: Number) = minus(Numbers(n))

  /**
   * Multiplies this and another numeric profile.
   *
   * @throws RealOpException if both profiles have non-zero rate at the same time.
   */
  operator fun times(other: SerialNumericOps<*, *>) = map2Values(other.toSerialLinear()) { l, r, i ->
    if (!l.isConstant() && !r.isConstant()) throw RealOpException("Cannot multiply two linear equations that are non-constant at the same time (at time ${i.start})")
    val shiftedRight = r.shiftInitialTime(l.initialTime)
    val newRate = l.rate * shiftedRight.initialValue + r.rate * l.initialValue
    LinearEquation(l.initialTime, l.initialValue * shiftedRight.initialValue, newRate)
  }

  /** Multiplies this by a constant number. */
  operator fun times(n: Number) = times(Numbers(n))

  /**
   * Calculates this divided by another numeric profile.
   *
   * @throws RealOpException if the divisor has a non-zero rate at any time that the dividend is defined.
   */
  operator fun div(other: SerialNumericOps<*, *>) = map2Values(other.toSerialLinear()) { l, r, i ->
    if (!r.isConstant()) throw RealOpException("Cannot divide by a non-piecewise-constant linear equation (at time ${i.start})")
    LinearEquation(l.initialTime, l.initialValue / r.initialValue, l.rate / r.initialValue)
  }

  /** Calculates this divided by a constant number. */
  operator fun div(n: Number) = div(Numbers(n))

  /**
   * Calculates this raised to the power of another numeric profile.
   *
   * @throws RealOpException if the exponent has a non-zero rate at any time that the base is defined,
   *                                 or if the base has a non-zero rate at any time that the exponent is defined and not
   *                                 either 0 or 1.
   */
  infix fun pow(exp: SerialNumericOps<*, *>) = map2Values(exp.toSerialLinear()) { l, r, i ->
    if (!r.isConstant()) throw RealOpException("Cannot apply a non-piecewise-constant exponent (at time ${i.start}")
    if (r.initialValue == 0.0) LinearEquation(1.0)
    else if (r.initialValue == 1.0) l
    else if (!l.isConstant()) throw RealOpException("Cannot apply an exponent to a non-piecewise-constant profile")
    else LinearEquation(l.initialValue.pow(r.initialValue))
  }

  /** Calculates this raised to the power of a constant number. */
  infix fun pow(n: Number) = pow(Numbers(n))

  /** Returns a [Booleans] that is true when this and another numeric profile are equal. */
  infix fun equalTo(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsEqualTo)
  /** Returns a [Booleans] that is true when this equals a constant number. */
  infix fun equalTo(n: Number) = equalTo(Numbers(n))

  /** Returns a [Booleans] that is true when this and another numeric profile are not equal. */
  infix fun notEqualTo(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsNotEqualTo)
  /** Returns a [Booleans] that is true when this does not equal a constant number. */
  infix fun notEqualTo(n: Number) = notEqualTo(Numbers(n))

  /** Returns a [Booleans] that is true when this is less than another numeric profile. */
  infix fun lessThan(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsLessThan)
  /** Returns a [Booleans] that is true when this is less than a constant number. */
  infix fun lessThan(n: Number) = lessThan(Numbers(n))

  /** Returns a [Booleans] that is true when this is less than or equal to another numeric profile. */
  infix fun lessThanOrEqualTo(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsLessThanOrEqualTo)
  /** Returns a [Booleans] that is true when this is less than or equal to a constant number. */
  infix fun lessThanOrEqualTo(n: Number) = lessThanOrEqualTo(Numbers(n))

  /** Returns a [Booleans] that is true when this is greater than another numeric profile. */
  infix fun greaterThan(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsGreaterThan)
  /** Returns a [Booleans] that is true when this is greater than a constant number. */
  infix fun greaterThan(n: Number) = greaterThan(Numbers(n))

  /** Returns a [Booleans] that is true when this is greater than or equal to another numeric profile. */
  infix fun greaterThanOrEqualTo(other: SerialNumericOps<*, *>) = inequalityHelper(other, LinearEquation::intervalsGreaterThanOrEqualTo)
  /** Returns a [Booleans] that is true when this is greater than or equal to a constant number. */
  infix fun greaterThanOrEqualTo(n: Number) = greaterThanOrEqualTo(Numbers(n))

  private fun inequalityHelper(other: SerialNumericOps<*, *>, f: LinearEquation.(LinearEquation) -> Booleans) =
      flatMap2Values(::Booleans, other.toSerialLinear()) { l, r, _ -> l.f(r) }


  override fun changes() =
      unsafeOperate(::Booleans) { opts ->
        val bounds = opts.bounds
        var previous: Segment<LinearEquation>? = null
        val result = collect(CollectOptions(bounds, false)).flatMap { currentSegment: Segment<LinearEquation> ->
          val currentInterval = currentSegment.interval
          val leftEdge = if (
              previous !== null &&
              previous!!.interval.compareEndToStart(currentInterval) == 0 &&
              currentInterval.includesStart()
          ) {
            previous!!.value.valueAt(currentInterval.start) == currentSegment.value.valueAt(currentInterval.start)
          } else if (currentInterval.compareStarts(bounds) == 0) {
            currentSegment.value.rate != 0.0
          } else {
            null
          }
          previous = currentSegment
          listOfNotNull(
              Segment(Interval.at(currentInterval.start), leftEdge).transpose(),
              Segment(Interval.between(currentInterval.start, currentInterval.end, Interval.Inclusivity.Exclusive), currentSegment.value.rate != 0.0)
          )
        }
        truncateList(result, opts)
      }

  /**
   * Returns a [Booleans] that is true whenever this discontinuously transitions between
   * a specific pair of values, and false or gap everywhere else.
   */
  fun transitions(from: Double, to: Double) = detectEdges(NullBinaryOperation.cases(
      { l, i -> if (l.valueAt(i.start) == from) null else false },
      { r, i -> if (r.valueAt(i.start) == to) null else false },
      { l, r, i -> l.valueAt(i.start) == from && r.valueAt(i.start) == to }
  ))

  override fun shiftedDifference(range: Duration) = shift(range.negate()).minus(this)

  /**
   * An exception for linear profile operations; usually thrown in contexts that
   * require one or more of the operands to be piecewise constant.
   */
  class RealOpException(message: String): Exception(message)

  /***/ companion object {
    /**
     * Converts a list of serialized value segments into a real profile; for use with [gov.nasa.jpl.aerie.timeline.plan.Plan.resource].
     *
     * Accepts either a map with the form `{initial: number, rate: number}`, or just a plain number for piecewise constant profiles.
     * While plain numbers are acceptable and will be converted to a [LinearEquation] without warning, consider using [Numbers]
     * to keep the precision.
     */
    @JvmStatic fun deserialize(list: List<Segment<SerializedValue>>): Real {
      val converted: List<Segment<LinearEquation>> = list.map { s ->
        s.value.asReal().getOrNull()?.let { return@map s.withNewValue(LinearEquation(it)) }
        val map = s.value.asMap().orElseThrow { RealDeserializeException("value was not a map or plain number: $s") }
        val initialValue = (map["initial"]
            ?: throw RealDeserializeException("initial value not found in map"))
            .asReal().orElseThrow { RealDeserializeException("initial value was not a double") }
        val rate = (map["rate"] ?: throw RealDeserializeException("rate not found in map"))
            .asReal().orElseThrow { RealDeserializeException("rate was not a double") }
        s.withNewValue(LinearEquation(s.interval.start, initialValue, rate))
      }
      return Real(converted)
    }

    /***/ class RealDeserializeException(message: String): Exception(message)
  }
}
